man/alarmtool.8
.TH ALARMTOOL 8 "Oct  3, 2006" "Nokia"

.SH NAME
alarmtool \- tool to test event daemon functionality

.SH SYNOPSIS
.B alarmtool
[\fIOPTION\fP]...

.SH DESCRIPTION
.B alarmtool
provides means to add different alarm events to queue of
.B alarmd.

.SH OPTIONS
.TP
\fB\-a\fP, \fB\-\-add\fP
Adds an event to the queue, see section
.B ADDING EVENTS.
.TP
\fB\-r\fP \fIeventid\fP, \fB\-\-remove\fP \fIeventid\fP
Removes event from the queue. The id supplied should be as got from
adding or querying events.
.TP
\fB\-q\fP, \fB\-\-query\fP
Queries events on the queue, see section 
.B QUERYING EVENTS.
.TP
\fB\-g\fP \fIeventid\fP, \fB\-\-get\fP \fIeventid\fP
Gets info about an event. The id supplied should be as got from adding or
querying events.

.SH ADDING EVENTS
When adding events, you need to specify certain details about the event.
On successful addition, the id of the added event is printed to standard
output. The details of the event should be set with following flags:
.TP
\fB\-t\fP \fItime\fP, \fB\-\-time\fP \fItime\fP
The time the event should launch, in seconds since epoch.
.TP
\fB\-R\fP \fIinterval\fP, \fB\-\-recurr\fP \fIinterval\fP
The time interval between two successive runs of the event in minutes.
Default is no repeating.
.TP
\fB\-C\fP \fIn\fP, \fB\-\-recurr-count\fP \fIn\fP
How many times should the event be repeated, \-1 for infinite.
.TP
\fB\-i\fP \fIiface\fP, \fB\-\-dbus-interface\fP \fIiface\fP
Interface for dbus action of the event.
.TP
\fB\-s\fP \fIservice\fP, \fB\-\-dbus-service\fP \fIservice\fP
Service to be called via DBus. If this is specified, dbus action will be a
method call.
.TP
\fB\-p\fP \fIpath\fP, \fB\-\-dbus-path\fP \fIpath\fP
Path for dbus action of the event.
.TP
\fB\-n\fP \fIname\fP, \fB\-\-dbus-name\fP \fIname\fP
Name of dbus method/signal.
.TP
\fB\-y\fP, \fB\-\-dbus-system-bus\fP
If used, the dbus action will be done over system bus, otherwise session
bus is used.
.TP
\fB\-m\fP \fImessage\fP, \fB\-\-dialog-message\fP \fImessage\fP
Message to show in the alarm dialog.
.TP
\fB\-c\fP \fIicon\fP, \fB\-\-dialog-icon\fP \fIicon\fP
Icon name or path to show in the alarm dialog.
.TP
\fB\-u\fP \fIsound\fP, \fB\-\-dialog-sound\fP \fIsound\fP
Path to a sound file to play when the alarm dialog is shown.
.TP
\fB\-e\fP \fIcommand\fP, \fB\-\-exec\fP \fIcommand\fP
Command to execute when the event is run. (You can only have DBus or exec,
not both).
.TP
\fB\-d\fP, \fB\-\-no-dialog\fP
If used, the event will not show a dialog when run.
.TP
\fB\-z\fP, \fB\-\-no-snooze\fP
If used, the snooze button in events alarm dialog is disabled.
.TP
\fB\-o\fP, \fB\-\-boot\fP
If used, the event will wake the device if powered off.
.TP
\fB\-b\fP, \fB\-\-show-icon\fP
If used, a alarm indicator will be shown in statusbar for the lifetime of the
event.
.TP
\fB\-r\fP, \fB\-\-run-delayed\fP
If used, the event will be run even if the time was missed due to powered
off device, or jumping over the set time. Otherwise the event will be removed
silently.
.TP
\fB\-w\fP, \fB\-\-connected\fP
If used, the event will be launched only when there is an internet connection.
If no connection is available at that time,
.TP
\fB\-a\fP, \fB\-\-activation\fP
If used, the dbus method call will use activation flag.
.TP
\fB\-P\fP, \fB\-\-postpone-delayed\fP
If used, the event will be moved forward to next day, if it is delayed for
more than one day.
.TP
\fB\-B\fP, \fB\-\-back-reschedule\fP
If used, a recurring alarm will be forced to be within recurr interval from
current time, and thus can be moved backwards, if device time is changed.
.B alarmd(8)
will wait for a connection and launch the event then.

.SH QUERYING EVENTS
The events on the queue can be queried based on their time and flags. The
list of ids for the events is printed to standard output. The default for
flags is not to care about them. If a flag and it's negative are both
specified, the last one is used. The details to query are set using following
flags:
.TP
\fB\-f\fP \fItime\fP, \fB\-\-first-time\fP \fItime\fP
Query events set to time greater than this. In seconds since epoch.
.TP
\fB\-l\fP \fItime\fP, \fB\-\-last-time\fP \fItime\fP
Query events set to time less than this. In seconds since epoch.
.TP
\fB\-Y\fP, \fB\-\-dbus-session-bus\fP
Query events that use session bus for dbus actions.
.TP
\fB\-y\fP, \fB\-\-dbus-system-bus\fP
Query events that use system bus for dbus actions.
.TP
\fB\-D\fP, \fB\-\-dialog\fP
Query events that use the alarm dialog.
.TP
\fB\-d\fP, \fB\-\-no-dialog\fP
Query events that do not use the alarm dialog.
.TP
\fB\-Z\fP, \fB\-\-snooze\fP
Query events that allow snooze in the alarm dialog.
.TP
\fB\-z\fP, \fB\-\-no-snooze\fP
Query events that do not allow snooze in the alarm dialog.
.TP
\fB\-o\fP, \fB\-\-boot\fP
Query events that power the device on, if powered off.
.TP
\fB\-O\fP, \fB\-\-no-boot\fP
Query events that power the device on, if powered off.
.TP
\fB\-b\fP, \fB\-\-show-icon\fP
Query events that show icon in the status bar.
.TP
\fB\-B\fP, \fB\-\-no-show-icon\fP
Query events that do not show icon in the status bar.
.TP
\fB\-r\fP, \fB\-\-run-delayed\fP
Query events that are run afterwards if missed due to power off or time
change.
.TP
\fB\-R\fP, \fB\-\-no-run-delayed\fP
Query events that are not run afterwards if missed due to power off or time
change.
.TP
\fB\-c\fP, \fB\-\-connected\fP
Query events that require an internet connection to be run.
.TP
\fB\-C\fP, \fB\-\-no-connected\fP
Query events that do net require an internet connection to be run.
.TP
\fB\-a\fP, \fB\-\-activation\fP
Query events that use DBus activation.
.TP
\fB\-A\fP, \fB\-\-no-activation\fP
Query events that do not use DBus activation.
.TP
\fB\-p\fP, \fB\-\-postpone-delayed\fP
Query events that are postponed if missed.
.TP
\fB\-P\fP, \fB\-\-no-postpne-dlayed\fP
Query events that are not postponed if missed.
.TP
\fB\-e\fP, \fB\-\-back-reschedule\fP
Query events that are moved backwards if time is changed backwards.
.TP
\fB\-E\fP, \fB\-\-no-back-reschedule\fP
Query events that are not moved backwards if time is changed backwards.

.SH RETURN VALUE
If an action was successful, the return value will be 0. If there was error
in input 1 is returned, and if an event specified could not be found, \-1.

.SH SEE ALSO
.BR alarmd (8)

.SH HISTORY
Oct  3 2006: Initial version of this manual page.

.SH COPYRIGHT
Copyright \(co 2006 Nokia Corporation.

This is free software; see /usr/share/common-licenses/LGPL-2.1 for license
conditions.  There is NO  warranty;  not even for MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE.


debian/changelog
alarmd (0.4.0) unstable; urgency=low

  * Version bump.
  * Reduce writes of data files. Fixes: NB#47023
  * Fix syntax errors in manpages.
  * Add debug logging.

 -- Santtu Lakkala <ext-santtu.1.lakkala@nokia.com>  Mon, 13 Nov 2006 13:19:44 +0200

alarmd (0.2.28) unstable; urgency=low

  * Create debug package too. Fixes: NB#46635

 -- Santtu Lakkala <ext-santtu.1.lakkala@nokia.com>  Wed,  8 Nov 2006 16:19:29 +0200

alarmd (0.2.27) unstable; urgency=low

  * Add NULL check. Fixes: NB#46535

 -- Santtu Lakkala <ext-santtu.1.lakkala@nokia.com>  Mon,  6 Nov 2006 15:08:49 +0200

alarmd (0.2.26) unstable; urgency=low

  * Fix DBus service started match string.

 -- Santtu Lakkala <ext-santtu.1.lakkala@nokia.com>  Wed, 25 Oct 2006 17:34:44 +0300

alarmd (0.2.25) unstable; urgency=low

  * Add more detailed match for service waiting. Fixes: NB#44955
  * Very minor cleanups.

 -- Santtu Lakkala <ext-santtu.1.lakkala@nokia.com>  Wed, 25 Oct 2006 13:36:04 +0300

alarmd (0.2.24) unstable; urgency=low

  * Minor cleanups. Fixes: NB#43912

 -- Santtu Lakkala <ext-santtu.1.lakkala@nokia.com>  Mon, 23 Oct 2006 15:28:14 +0300

alarmd (0.2.23) unstable; urgency=low

  * Various cleanups. Fixes: NB#43037

 -- Santtu Lakkala <ext-santtu.1.lakkala@nokia.com>  Wed, 11 Oct 2006 17:41:19 +0300

alarmd (0.2.22) unstable; urgency=low

  * Don't clear hw clock alarm on deinit.

 -- Santtu Lakkala <ext-santtu.1.lakkala@nokia.com>  Mon,  9 Oct 2006 15:56:25 +0300

alarmd (0.2.21) unstable; urgency=low

  * Launch new events -- not old ones -- when time changes.
  * Fix slist handling in AlarmdQueue time_changed method.
    Fixes: NB#42433
  * More manpages.
  * Fix segmentation fault on deinitialization.
  * Make timer plugins aware of initialization state. Fixes: NB#40727.

 -- Santtu Lakkala <ext-santtu.1.lakkala@nokia.com>  Mon,  9 Oct 2006 13:47:00 +0300

alarmd (0.2.20) unstable; urgency=low

  * Add clear used data script. Fixes: NB#41687

 -- Santtu Lakkala <ext-santtu.1.lakkala@nokia.com>  Wed,  4 Oct 2006 16:07:35 +0300

alarmd (0.2.19) unstable; urgency=low

  * Show powerup dialog also on snooze. Fixes: NB#35990
  * Add ALARM_EVENT_BACK_RESCHEDULE flag.
  * Add manpages.

 -- Santtu Lakkala <ext-santtu.1.lakkala@nokia.com>  Wed,  4 Oct 2006 10:25:54 +0300

alarmd (0.2.18) unstable; urgency=low

  * Add ALARM_EVENT_POSTPONE_DELAYED flag.
  * Move flags from actiondialog to action.

 -- Santtu Lakkala <ext-santtu.1.lakkala@nokia.com>  Wed, 27 Sep 2006 15:18:00 +0300

alarmd (0.2.17) unstable; urgency=low

  * Use own retutime wrappers instead of osso-clock.
  * Separate more rpc calls to own files.
  * Use new osso-systemui-dbus-dev.

 -- Santtu Lakkala <ext-santtu.1.lakkala@nokia.com>  Thu, 21 Sep 2006 16:30:24 +0300

alarmd (0.2.16) unstable; urgency=low

  * Add alarm_unescape_string to libalarm.
  * Iterate all events on query, including pending ones.
  * Send 'off' messages to mce more verbosely. Fixes: NB#40414.
  * Save queue if it changes during startup.
  * When connected event is launched, wait for connection if not
    connected. Fixes: NB#30579.
  * Do some logging on error cases.
  * Fix delayed event handling in timer plugins. Fixes: NB#40680.
  * Fix dequeue event handling.

 -- Santtu Lakkala <ext-santtu.1.lakkala@nokia.com>  Mon, 18 Sep 2006 16:23:01 +0300

alarmd (0.2.15) unstable; urgency=low

  * Send original alarm time to systemui.
  * Add alarm_escape_string to libalarm.

 -- Santtu Lakkala <ext-santtu.1.lakkala@nokia.com>  Wed, 13 Sep 2006 10:44:22 +0300

alarmd (0.2.14) unstable; urgency=low

  * Improve recurring alarm postponing algorithm.
  * Save queue when snooze time gets changed.
  * Fix snooze on old alarms. Fixes: NB#39451.
  * Tell systemui to replay sound. Fixes: NB#36239.
  * Acknowledge pending alarms on shutdown. Fixes: NB#38636.
  * Include configuration file for osso-backup. Fixes: NB#36536.

 -- Santtu Lakkala <ext-santtu.1.lakkala@nokia.com>  Mon, 11 Sep 2006 13:23:45 +0300

alarmd (0.2.13) unstable; urgency=low

  * Improve mce communication.
  * Use powerup dialog in dbus and exec action types too. Fixes:
    NB#35990.

 -- Santtu Lakkala <ext-santtu.1.lakkala@nokia.com>  Thu,  7 Sep 2006 09:59:33 +0300

alarmd (0.2.12) unstable; urgency=low

  * Remove .la and .a files from package. Fixes: NB#38467.
  * Call mce with alarm dialog visibility status. Fixes: NB#39047.
  * Add NULL terminator to g_object_get call. Fixes: NB#37662.
  * Call timer plugin time_changed methods when time changes. Fixes:
    NB#38600.

 -- Santtu Lakkala <ext-santtu.1.lakkala@nokia.com>  Tue,  5 Sep 2006 13:45:44 +0300

alarmd (0.2.11) unstable; urgency=low

  * Fix storing the snooze time.

 -- Santtu Lakkala <ext-santtu.1.lakkala@nokia.com>  Wed,  2 Aug 2006 16:27:57 +0300

alarmd (0.2.10) unstable; urgency=low

  * Added set_get default snooze functions and snoozed field to
    libalarm/alarm_event_t sctruct and changed the implementation
    accordingly. Fixes: NB#35308.
  * Changed to signleton dbus bus handling. Fixes: NB#34449.
  * Fixed queue to continue working even when an alarm dialog is
    showing.
  * Added timer-interface.h to libalarm-dev package.
  * Added error reporting to libalarm.
  * Added activation flag to events.

 -- Santtu Lakkala <ext-santtu.1.lakkala@nokia.com>  Wed,  2 Aug 2006 10:41:13 +0300

alarmd (0.2.9) unstable; urgency=low

  * Only use retu to get device up, gtimeouts for normal timing for
    finer granularity. Fixes: NB#35183.
  * Act on time changes that pass over the alarm time. Fixes: NB#35244.
  * Change loading order of timers and queue in order to get queued
    items on right queue.
  * Fix dbus bus handling in statusbar subsystem.

 -- Santtu Lakkala <ext-santtu.1.lakkala@nokia.com>  Tue, 25 Jul 2006 14:20:35 +0300

alarmd (0.2.8) unstable; urgency=low

  * Add alarm_dbus.h to libalarm-dev.

 -- Santtu Lakkala <ext-santtu.1.lakkala@nokia.com>  Fri, 14 Jul 2006 14:05:13 +0300

alarmd (0.2.7) unstable; urgency=low


alarmtool.c
/**
 * This file is part of alarmd
 *
 * Contact Person: David Weinehall <david.weinehall@nokia.com>
 *
 * Copyright (C) 2006 Nokia Corporation
 * alarmd and libalarm are free software; you can redistribute them
 * and/or modify them under the terms of the GNU Lesser General Public
 * License version 2.1 as published by the Free Software Foundation.
 *
 * alarmd and libalarm are distributed in the hope that they will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this software; if not, write to the Free
 * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 * 02110-1301 USA
 */

#include <getopt.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include "include/alarm_event.h"

extern char *optarg;
extern int optind;

static int tool_add(int argc, char **argv);
static int tool_remove(const char *arg);
static int tool_list(int argc, char **argv);
static int tool_get(const char *arg);
static void usage(const char *name);

int main(int argc, char **argv)
{
	int optc;
	int opt_index;

	const char optline[] = "ar:qg:";

	struct option const options_top[] = {
		{ "add", 0, 0, 'a' },
		{ "remove", 1, 0, 'r' },
		{ "query", 0, 0, 'q' },
		{ "get", 1, 0, 'g' },
		{ 0, 0, 0, 0 }
	};

	while ((optc = getopt_long(argc, argv, optline,
					options_top, &opt_index)) != -1) {
		switch (optc) {
		case 'a':
			return tool_add(argc, argv);
			break;
		case 'r':
			return tool_remove(optarg);
			break;
		case 'q':
			return tool_list(argc, argv);
			break;
		case 'g':
			return tool_get(optarg);
			break;
		default:
			usage(argv[0]);
			return 1;
		}
	}

	usage(argv[0]);
	return 1;

	return 0;
}

static int tool_add(int argc, char **argv)
{
	int optc;
	int opt_index;
	alarm_event_t new_event;
	long event_id;

	memset(&new_event, 0, sizeof(new_event));

	struct option const options_add[] = {
		{ "time", 1, 0, 't' },
		{ "recurr", 1, 0, 'R' },
		{ "recurr-count", 1, 0, 'C' },
		{ "dbus-interface", 1, 0, 'i' },
		{ "dbus-service", 1, 0, 's' },
		{ "dbus-path", 1, 0, 'p' },
		{ "dbus-name", 1, 0, 'n' },
		{ "dbus-system-bus", 0, 0, 'y' },
		{ "dialog-message", 1, 0, 'm' },
		{ "dialog-icon", 1, 0, 'c' },
		{ "dialog-sound", 1, 0, 'u' },
		{ "exec", 1, 0, 'e' },
		{ "no-dialog", 0, 0, 'd' },
		{ "no-snooze", 0, 0, 'z' },
		{ "boot", 0, 0, 'o' },
		{ "show-icon", 0, 0, 'b' },
		{ "run-delayed", 0, 0, 'r' },
		{ "connected", 0, 0, 'w' },
		{ "activation", 0, 0, 'a' },
		{ "postpone-delayed", 0, 0, 'P' },
		{ "back-reschedule", 0, 0, 'B' },
		{ 0, 0, 0, 0 }
	};

	const char optline[] = "t:R:C:i:s:p:n:ym:c:u:e:dzobrwaPB";

	while ((optc = getopt_long(argc, argv, optline,
					options_add, &opt_index)) != -1) {
		switch (optc) {
		case 'm':
			if (new_event.message) {
				free(new_event.message);
			}
			new_event.message = strdup(optarg);
			break;
		case 'c':
			if (new_event.icon) {
				free(new_event.icon);
			}
			new_event.icon = strdup(optarg);
			break;
		case 'u':
			if (new_event.sound) {
				free(new_event.sound);
			}
			new_event.sound = strdup(optarg);
			break;
		case 't':
			new_event.alarm_time = atol(optarg);
			break;
		case 'R':
			new_event.recurrence = atoi(optarg);
			break;
		case 'C':
			new_event.recurrence_count = atoi(optarg);
			break;
		case 'i':
			if (new_event.dbus_interface) {
				free(new_event.dbus_interface);
			}
			new_event.dbus_interface = strdup(optarg);
			break;
		case 's':
			if (new_event.dbus_service) {
				free(new_event.dbus_service);
			}
			new_event.dbus_service = strdup(optarg);
			break;
		case 'p':
			if (new_event.dbus_path) {
				free(new_event.dbus_path);
			}
			new_event.dbus_path = strdup(optarg);
			break;
		case 'n':
			if (new_event.dbus_name) {
				free(new_event.dbus_name);
			}
			new_event.dbus_name = strdup(optarg);
			break;
		case 'e':
			if (new_event.exec_name) {
				free(new_event.exec_name);
			}
			new_event.exec_name = strdup(optarg);
			break;
		case 'y':
			new_event.flags |= ALARM_EVENT_SYSTEM;
			break;
		case 'd':
			new_event.flags |= ALARM_EVENT_NO_DIALOG;
			break;
		case 'z':
			new_event.flags |= ALARM_EVENT_NO_SNOOZE;
			break;
		case 'o':
			new_event.flags |= ALARM_EVENT_BOOT;
			break;
		case 'b':
			new_event.flags |= ALARM_EVENT_SHOW_ICON;
			break;
		case 'r':
			new_event.flags |= ALARM_EVENT_RUN_DELAYED;
			break;
		case 'w':
			new_event.flags |= ALARM_EVENT_CONNECTED;
			break;
		case 'a':
			new_event.flags |= ALARM_EVENT_ACTIVATION;
			break;
		case 'P':
			new_event.flags |= ALARM_EVENT_POSTPONE_DELAYED;
			break;
		case 'B':
			new_event.flags |= ALARM_EVENT_BACK_RESCHEDULE;
			break;
		default:
			fprintf(stderr, "%s: got %c\n", __FUNCTION__, optc);
			break;
		}
	}
	event_id = alarm_event_add(&new_event);
	if (event_id) {
		printf("Event added with id %ld\n", event_id);
		return 0;
	}
	else {
		fprintf(stderr, "Adding event failed.\n");
		return 1;
	}
	return 0;
}

static int tool_list(int argc, char **argv)
{
	int optc;
	int opt_index;

	int32_t flag_mask = 0;
	int32_t flags = 0;

	time_t start_time = 0;
	time_t end_time = ~0; 

	long *events;

	if (end_time < 0) {
		end_time--;
	}

	struct option const options_list[] = {
		{ "first-time", 1, 0, 'f' },
		{ "last-time", 1, 0, 'l' },
		{ "dbus-system-bus", 0, 0, 'y' },
		{ "no-dialog", 0, 0, 'd' },
		{ "no-snooze", 0, 0, 'z' },
		{ "boot", 0, 0, 'o' },
		{ "show-icon", 0, 0, 'b' },
		{ "run-delayed", 0, 0, 'r' },
		{ "dbus-session-bus", 0, 0, 'Y' },
		{ "dialog", 0, 0, 'D' },
		{ "snooze", 0, 0, 'Z' },
		{ "no-boot", 0, 0, 'O' },
		{ "no-show-icon", 0, 0, 'B' },
		{ "no-run-delayed", 0, 0, 'R' },
		{ "connected", 0, 0, 'c' },
		{ "no-connected", 0, 0, 'C' },
		{ "activation", 0, 0, 'a' },
		{ "no-activation", 0, 0, 'A' },
		{ "postpone-delayed", 0, 0, 'p' },
		{ "no-postpne-dlayed", 0, 0, 'P' },
		{ "back-reschedule", 0, 0, 'e' },
		{ "no-back-reschedule", 0, 0, 'E' },
		{ 0, 0, 0, 0 }
	};

	const char optline[] = "f:l:ydzobrYDZOBRcCaApPeE";

	while ((optc = getopt_long(argc, argv, optline,
					options_list, &opt_index)) != -1) {
		switch (optc) {
		case 'f':
			start_time = atol(optarg);
			break;
		case 'l':
			end_time = atol(optarg);
			break;
		case 'y':
			flags |= ALARM_EVENT_SYSTEM;
			flag_mask |= ALARM_EVENT_SYSTEM;
			break;
		case 'd':
			flags |= ALARM_EVENT_NO_DIALOG;
			flag_mask |= ALARM_EVENT_NO_DIALOG;
			break;
		case 'z':
			flags |= ALARM_EVENT_NO_SNOOZE;
			flag_mask |= ALARM_EVENT_NO_SNOOZE;
			break;
		case 'o':
			flags |= ALARM_EVENT_BOOT;
			flag_mask |= ALARM_EVENT_BOOT;
			break;
		case 'b':
			flags |= ALARM_EVENT_SHOW_ICON;
			flag_mask |= ALARM_EVENT_SHOW_ICON;
			break;

		case 'r':
			flags |= ALARM_EVENT_RUN_DELAYED;
			flag_mask |= ALARM_EVENT_RUN_DELAYED;
			break;
		case 'Y':
			flags &= ~ALARM_EVENT_SYSTEM;
			flag_mask |= ALARM_EVENT_SYSTEM;
			break;
		case 'D':
			flags &= ~ALARM_EVENT_NO_DIALOG;
			flag_mask |= ALARM_EVENT_NO_DIALOG;
			break;
		case 'Z':
			flags &= ~ALARM_EVENT_NO_SNOOZE;
			flag_mask |= ALARM_EVENT_NO_SNOOZE;
			break;
		case 'O':
			flags &= ~ALARM_EVENT_BOOT;
			flag_mask |= ALARM_EVENT_BOOT;
			break;
		case 'B':
			flags &= ~ALARM_EVENT_SHOW_ICON;
			flag_mask |= ALARM_EVENT_SHOW_ICON;
			break;
		case 'R':
			flags &= ~ALARM_EVENT_RUN_DELAYED;
			flag_mask |= ALARM_EVENT_RUN_DELAYED;
			break;
		case 'c':
			flags |= ALARM_EVENT_CONNECTED;
			flag_mask |= ALARM_EVENT_CONNECTED;
			break;
		case 'C':
			flags &= ~ALARM_EVENT_CONNECTED;
			flag_mask |= ALARM_EVENT_CONNECTED;
			break;
		case 'a':
			flags |= ALARM_EVENT_ACTIVATION;
			flag_mask |= ALARM_EVENT_ACTIVATION;
			break;
		case 'A':
			flags &= ~ALARM_EVENT_ACTIVATION;
			flag_mask |= ALARM_EVENT_ACTIVATION;
			break;
		case 'p':
			flags |= ALARM_EVENT_POSTPONE_DELAYED;
			flag_mask |= ALARM_EVENT_POSTPONE_DELAYED;
			break;
		case 'P':
			flags &= ~ALARM_EVENT_POSTPONE_DELAYED;
			flag_mask |= ALARM_EVENT_POSTPONE_DELAYED;
			break;
		case 'e':
			flags |= ALARM_EVENT_BACK_RESCHEDULE;
			flag_mask |= ALARM_EVENT_BACK_RESCHEDULE;
			break;
		case 'E':
			flags &= ~ALARM_EVENT_BACK_RESCHEDULE;
			flag_mask |= ALARM_EVENT_BACK_RESCHEDULE;
			break;
		default:
			return 1;
			break;
		}
	}

	events = alarm_event_query(start_time, end_time, flag_mask, flags);

	if (events == NULL) {
		fprintf(stderr, "Error while querying the events.\n");
		return -1;
	} else {
		if (*events == 0) {
			printf("No events.\n");
		} else {
			unsigned int i = 0;

			printf("Events found:\n");
			for (i = 0; events[i] != 0; i++) {
				printf("\t%ld\n", events[i]);
			}
		}
		free(events);
	}

	return 0;
}

static int tool_remove(const char *arg)
{
	char* err;
	cookie_t event_id=strtol(arg,&err,10);

	if (*err != '\0') {
		// Invalid cookie
		fprintf(stderr,"Invalid event id \"%s\"\n",arg);
		return 1;
	}

	if (alarm_event_del(event_id)) {
		// event removed
		return 0;
	} else {
		fprintf(stderr,"Error while removing event \"%s\"\n",arg);
		// Error while removing the event
	}
	return -1;
}

static int tool_get(const char *arg)
{
	alarm_event_t* event;
	cookie_t event_id;
	char* err;
	fprintf(stderr, "%s(%s)\n", __FUNCTION__, arg);

	event_id=strtol(arg,&err,10);

	if (*err != '\0') {
		fprintf(stderr,"Invalid event id \"%s\"\n",arg);
		// Invalid cookie
		return 1;
	}

	event = alarm_event_get(event_id);

	if (event==NULL) {
		fprintf(stderr, "Can't get event \"%s\"\n", arg);
		// Error

		return -1;
	} else {
		const char *system, *dialog, *snooze, *boot, *actdead,
			*icon, *delayed, *connected, *activation,
			*postpone, *breschedule;

		if (event->flags & ALARM_EVENT_SYSTEM) {
			system = "System ";
		} else {
			system = "";
		}

		if (event->flags & ALARM_EVENT_NO_DIALOG) {
			dialog = "No_dialog ";
		} else {
			dialog = "";
		}

		if (event->flags & ALARM_EVENT_NO_SNOOZE) {
			snooze = "No_snooze ";
		} else {
			snooze = "";
		}

		if (event->flags & ALARM_EVENT_BOOT) {
			boot = "Boot ";
		} else {
			boot = "";
		}	  

		if (event->flags & ALARM_EVENT_ACTDEAD) {
			actdead = "Boot_Actdead ";
		} else {
			actdead = "";
		}

		if (event->flags & ALARM_EVENT_SHOW_ICON) {
			icon = "Icon ";
		} else {
			icon = "";
		}

		if (event->flags & ALARM_EVENT_RUN_DELAYED) {
			delayed = "No_dialog ";
		} else {
			delayed = "";
		}

		if (event->flags & ALARM_EVENT_CONNECTED) {
			connected = "Connected ";
		} else {
			connected = "";
		}

		if (event->flags & ALARM_EVENT_ACTIVATION) {
			activation = "Activation ";
		} else {
			activation = "";
		}

		if (event->flags & ALARM_EVENT_POSTPONE_DELAYED) {
			postpone = "Postpone ";
		} else {
			postpone = "";
		}

		if (event->flags & ALARM_EVENT_BACK_RESCHEDULE) {
			breschedule = "Back_Reschedule ";
		} else {
			breschedule = "";
		}

		printf("\ttime\t: \"%u\"\n"
				"\tmgs\t: \"%s\"\n"
				"\ticon\t: \"%s\"\n"
				"\tsound\t: \"%s\"\n"
				"\tiface\t: \"%s\"\n"
				"\tservice\t: \"%s\"\n"
				"\tpath\t: \"%s\"\n"
				"\tname\t: \"%s\"\n"
				"\texec-name\t: \"%s\"\n"
				"\tflags\t: \"%s%s%s%s%s%s%s%s%s%s%s\"\n"
				,
				(unsigned)event->alarm_time,
				event->message,
				event->icon,
				event->sound,
				event->dbus_interface,
				event->dbus_service,
				event->dbus_path,
				event->dbus_name,
				event->exec_name,
				dialog, snooze, system, boot, actdead, icon,
				delayed, connected, activation, postpone,
				breschedule
		      );

	}
	return 0;
}

static void usage(const char *name)
{
	printf("Usage:\n"
			"\t%s -a [-tispnymcuedzobr] [option params]\n"
			"\t%s -r <cookie>\n"
			"\t%s -l [-ydzobrYDZOBR]\n"
			"\t%s -g <cookie>\n\n"
			"\t-a, --add\tAdd event to queue\n"
			"\t-t, --time\n", name, name, name, name);
}
